swift/apple/FirezoneNetworkExtension/SessionManager.swift

```diff
@@ -0,0 +1,250 @@
+  }
+
+  /// The main event loop that owns the Session lifecycle.
+  ///
+  /// This uses withTaskGroup to multiplex between:
+  /// 1. Event polling from Rust (nextEvent)
+  /// 2. Command handling from Swift (command channel)
+  ///
+  /// When either task completes, both are cancelled and the session is cleaned up.
+  /// This ensures the Session's Drop is called on the Rust side.
+  private func runEventLoop(session: Session) async {
+    Log.log("SessionManager: Starting event loop")
+
+    // Multiplex between commands and events
+    await withTaskGroup(of: Void.self) { group in
+      // Event polling task - polls Rust for events
+      group.addTask { [weak self] in
+        guard let self = self else { return }
+
+        var eventCount = 0
+        var pollAttempts = 0
+
+        while !Task.isCancelled {
+          pollAttempts += 1
+
+          do {
+            // Poll for next event from Rust
+            if let event = try await session.nextEvent() {
+              eventCount += 1
+              Log.log(
+                "SessionManager: Event received (\(eventCount)): \(String(describing: event))")
+              await self.eventHandler(event)
+            } else {
+              // No event returned - session has ended
+              Log.log("SessionManager: Event stream ended, exiting event loop")
+              break
+            }
+          } catch {
+            Log.error(error)
+            Log.log("SessionManager: Error in event polling, exiting event loop")
+            break
+          }
+
+          // Periodic health check logging
+          if pollAttempts % Self.eventPollingLogInterval == 0 {
+            Log.log("SessionManager: Event polling active - \(eventCount) events processed")
+          }
+        }
+
+        Log.log("SessionManager: Event polling finished after \(eventCount) events")
+      }
+
+      // Command handling task - processes commands from Swift
+      group.addTask { [weak self] in
+        guard let self = self else { return }
+
+        for await command in self.commandStream {
+          await self.handleCommand(command, session: session)
+
+          // Exit loop if disconnect command
+          if case .disconnect = command {
+            Log.log("SessionManager: Disconnect command received, exiting command loop")
+            break
+          }
+        }
+
+        Log.log("SessionManager: Command handling finished")
+      }
+
+      // Wait for first task to complete, then cancel all
+      _ = await group.next()
+      Log.log("SessionManager: One task completed, cancelling event loop")
+      group.cancelAll()
+    }
+
+    // Cleanup when event loop exits
+    // Assigning to `nil` will invoke `Drop` on the Rust side
+    // Do NOT call disconnect() explicitly - let Drop handle everything
+    Log.log("SessionManager: Event loop finished, cleaning up session")
+    self.session = nil
+    Log.log("SessionManager: Session cleaned up")
```

Comment:
This seems redundant and is the kind of stuff I'd like to avoid. If we "move" the session into the event-loop, breaking from both loops should reduce the reference count to zero and free all the memory, meaning we don't need to actually assign something to `nil` to free the last reference.